rules/common/src/main/kotlin/io/nlopez/compose/rules/ViewModelForwarding.kt

```diff
@@ -58,27 +63,82 @@ class ViewModelForwarding : ComposeKtVisitor {
-                    .map { callExpression }
-            }
-            .forEach { callExpression ->
-                emitter.report(callExpression, AvoidViewModelForwarding, false)
-            }
+        val checkedCallExpressions = HashSet<KtCallExpression>()
+        fun checkCallExpressions(
+            bodyExpression: KtBlockExpression?,
+            scopedParameter: String? = null,
+            usesItObjectRef: Boolean = false,
+        ) {
+            bodyExpression?.findChildrenByClass<KtCallExpression>()
+                ?.filterNot { it in checkedCallExpressions }
+                ?.forEach { callExpression ->
+                    checkedCallExpressions.add(callExpression)
+                    // We want now to see if these parameter names are used in any other calls to functions that start with
+                    // a capital letter (so, most likely, composables).
+                    if (callExpression.calleeExpression?.text?.first()?.isUpperCase() == true &&
+                        // Avoid LaunchedEffect/DisposableEffect/etc that can use the VM as a key
+                        !callExpression.isRestartableEffect &&
+                        // Avoid explicitly allowlisted Composable names
+                        allowedForwardingRegex?.let { callExpression.calleeExpression?.text?.matches(it) } != true
+                    ) {
+                        // Get VALUE_ARGUMENT that has a REFERENCE_EXPRESSION. This would map to `viewModel` in this example:
+                        // MyComposable(viewModel, ...)
+                        callExpression.valueArguments
+                            .mapNotNull { valueArgument ->
+                                when (val argumentExpression = valueArgument.getArgumentExpression()) {
+                                    is KtReferenceExpression, is KtThisExpression -> argumentExpression
+                                    else -> null
+                                }
+                            }
+                            .filter { reference ->
+                                reference.text in viewModelParameterNames ||
+                                    (
+                                        reference.text == "it" &&
+                                            scopedParameter in viewModelParameterNames && usesItObjectRef
+                                        ) ||
+                                    (
+                                        reference.text == "this" &&
+                                            scopedParameter in viewModelParameterNames && !usesItObjectRef
+                                        )
+                            }
+                            .forEach { _ ->
+                                emitter.report(callExpression, AvoidViewModelForwarding, false)
+                            }
+                    }
+
+                    val scopedFunctions = setOf("with", "apply", "run", "also", "let")
```

Comment:
these should be defined in a val in the companion object so they aren't instantiated multiple times